cranelift/codegen/src/isa/riscv64/inst/emit.rs

```diff
@@ -1193,15 +1193,39 @@ impl MachInstEmit for Inst {
                             .into_iter()
                             .for_each(|inst| inst.emit(&[], sink, emit_info, state));
                     }
-                    (AMode::Const(addr), None, _) => unimplemented!("LoadAddr: {:?}", addr),
-                    (AMode::Label(label), None, _) => unimplemented!("LoadAddr: {:?}", label),
+                    (AMode::Const(addr), None, _) => {
+                        // Get an address label for the constant and recurse.
+                        let label = sink.get_label_for_constant(addr);
+                        Inst::LoadAddr {
+                            rd,
+                            mem: AMode::Label(label),
+                        }
+                        .emit(&[], sink, emit_info, state);
+                    }
+                    (AMode::Label(label), None, _) => {
+                        // Get the current PC.
+                        sink.use_label_at_offset(sink.cur_offset(), label, LabelUse::PCRelHi20);
+                        let inst = Inst::Auipc {
+                            rd,
+                            imm: Imm20::from_bits(0),
+                        };
+                        inst.emit(&[], sink, emit_info, state);
+
+                        // Emit an add to the address with a relocation.
+                        // This later gets patched up with the correct offset.
+                        sink.use_label_at_offset(sink.cur_offset(), label, LabelUse::PCRelLo12I);
+                        Inst::AluRRImm12 {
+                            alu_op: AluOPRRI::Addi,
+                            rd,
+                            rs: rd.to_reg(),
+                            imm12: Imm12::zero(),
+                        }
+                        .emit(&[], sink, emit_info, state);
```

Comment:
This comment reminded me that we discussed something almost exactly like this a while ago. And I found it: 
https://github.com/bytecodealliance/wasmtime/pull/5951#issuecomment-1459129236

The `Call` Relocation in the JIT performs exactly the same arithmetic. And it is different from this one, we only offset the Lo12 by 4 not the Hi20, which makes sense.

I'm also going to try to reproduce that off-by-one with a test, I get really confused by this relocation math.

Reply:
I think I got this right, but I'm never confident when this sort of relocation math is involved.

We had 2 issues here. The first one you pointed out, only the Lo12 relocation should have the 4byte offset. The second one was the Hi20 relocation, that needs to get a `0x800` offset so that it skips to the next page as soon as the offset goes out of range for Lo12 since it is signed.